keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+Simple `printf` like logging is currently not enough for today’s distributed system. Thankfully since kubernetes inception the observability for cloud native applications evolved significantly. Unification and correlation of all telemetry verticals (logs, metrics, traces) has become an important feature of SaaS provides like Splunk, Datadog or Sumologic and with CNCF projects like OpenTelemetry starts to become standard. Many log storage backends like BigQuery, Elastic and CloudWatch Logs have matured allowing to efficiently process and analyze massive datasets of structured data. Such solutions rely heavily on consistency of metadata, which is required for efficient indexing and filtering. An example of projects that tackles this problem is Elastic Common Schema, which can be adapted by application independent of the language they are written in.
+
+All those changes have led to a situation when most of the logs produced by distributed applications is now not stored locally but instead is aggregated and sent to some sort of external storage system. This makes working with logs much easier and provides completely new opportunities for using logs for different kinds of quantitative analyzes but at the same time moves the focus for logs from being easy to read for humans to be easy to process and index by logging services.
+
+Because Kubernetes is a very technical solution on which lots of developers work on a daily basis we should not abandon completely developers centered approach for logging but instead add a way for Kubernetes users to choose between those two perspectives depending on the cluster purpose - if the given cluster is used for development or is it a production cluster. However this does not mean that the human-readable output format provided for developers needs to be exactly the same as the current klog format.
+
+Apart from being easy to ingest into logging services logs in order to be useful needs also to be searchable meaning that it should be easy for the user to find the most relevant information in them. Solutions like Elastic or BigQuery makes querying structured logs very easy. What they do not provide is the consistent schema to be used by those logs. Without it querying logs will be still possible but finding specific information might be much harder. Therefore apart from simply having some structure logs produced by applications should also follow a schema which would enforce consistent naming, consistent way of referencing the same objects and common set of metadata to be stored for each of those objects. With such schema in place building queries for logs becomes much easier.
+
+### Goals
+
+* Provide structured logging interface in K8s codebase
+* Standardize commonly used logging keys (e.g. referencing k8s object, http request fields)
+* Propose sound strategy for migration ensuring its completeness
+* Make implementation of context based solutions like logs correlation or tracing within Kubernetes easier.
```

Comment:
This seems like a distinct initiative. Also worthwhile (perhaps more-so), but distinct from structuring logs

Reply:
Yes, indeed it might be considered a distinctive initiative but as it would be a big task on its own and would require significant amount of work I do not believe that the benefits coming out of it would justify it. This changes if we can make those things much easier to implement as a kind of side effect of introducing structured logs.

Reply:
We decided to follow your suggestion and narrow the scope of this KEP. Thanks for your feedback